bundles/alpha.codegen/src/alpha/codegen/writeC/SystemConverter.xtend

```diff
@@ -0,0 +1,473 @@
+import static extension alpha.codegen.isl.PolynomialConverter.convert
+import static extension alpha.codegen.writeC.Common.getEvalName
+import static extension alpha.codegen.writeC.Common.getFlagName
+import static extension alpha.model.util.AlphaUtil.copyAE
+import static extension fr.irisa.cairn.jnimap.barvinok.BarvinokBindings.card
+
+/** Converts an Alpha system to the simplified C AST. */
+class SystemConverter {
+	/** The list of all variables allocated by the entry point function which need to be freed before returning. */
+	protected val ArrayList<String> allocatedVariables
+
+	/**
+	 * A counter for the number of statement macros that have been created.
+	 * This is used for determining the names said macros.
+	 */
+	var nextStatementId = 0
+	
+	/**
+	 * Flag to indicate that the C code being output should be compatible
+	 * with the wrapper produced by the older AlphaZ.
+	 */
+	protected val boolean oldAlphaZCompatible
+	
+	/** The instance of the program builder. */
+	protected val ProgramBuilder program
+	
+	/** The system being converted into a C program. */
+	protected val AlphaSystem system
+	
+	/** Protected constructor. */
+	protected new(AlphaSystem system, boolean oldAlphaZCompatible) {
+		this.allocatedVariables = newArrayList
+		this.oldAlphaZCompatible = oldAlphaZCompatible
+		this.program = ProgramBuilder.start
+		this.system = system
+	}
+	
+	/**
+	 * Converts an Alpha system to the simplified C AST.
+	 * Only supports systems with a single body.
+	 */
+	def static convert(AlphaSystem system) {
+		return convert(system, false)
+	}
+	
+	/**
+	 * Converts an Alpha system to the simplified C AST.
+	 * Only supports systems with a single body.
+	 * 
+	 * If requested, the code produced will aim for compatibility with
+	 * the older version of AlphaZ (although with no guarantees).
+	 * This means that the inputs to the entry point will assume
+	 * that memory for the system's inputs and outputs were allocated
+	 * via the bounding box method, as opposed to linearized memory.
+	 * 
+	 * Note: since the newer Alpha language doesn't have typing,
+	 * all indices are assumed to be of type "long",
+	 * and all data values are of type "float".
+	 */
+	def static convert(AlphaSystem system, boolean oldAlphaZCompatible) {
+		// Duplicate the system and then preprocess the duplicate
+		// by applying normalization and name standardization.
+		val duplicate = system.copyAE
+		Normalize.apply(duplicate)
+		StandardizeNames.apply(duplicate)
+		
+		return (new SystemConverter(duplicate, oldAlphaZCompatible)).convertSystem
+	}
+	
+	/**
+	 * Converts an Alpha system to the simplified C AST.
+	 * Only supports systems with a single body.
+	 */
+	def protected convertSystem() {
+		// Only systems with a single body are currently supported.
+		if (system.systemBodies.size != 1) {
+			throw new IllegalArgumentException("Only systems with one body are supported, but this system has " + system.systemBodies.size + ".")
+		}
```

Comment:
I don’t think external functions are supported. Could you just add a check & throw a “not yet supported”exception.

Reply:
I couldn't add one here, as the `AlphaSystem` class doesn't actually contain the external functions (those are in the `AlphaRoot` class, which I don't have easy access to). However, I added all the expressions that use external functions to the `ExprConverter` and explicitly threw exceptions there. This means external functions can be declared, but not used.

See commit da83c3f

Reply:
The `alpha.model.util.AlphaUtil.getContainerRoot` can be used to get the root from any node in the AST. This can’t be used here? Either way, the way you’re doing it that commit is fine by me. 

I also created issue #53 to track adding this in the future.